### DISCUSSION TITLE
Classes and dependency inversion principle

### DISCUSSION THREAD

**Discussion by mapeveri** [fastapi/fastapi#9008] [Questions] [answered]
I have an app using hexagonal architecture and i'm trying to apply the dependency inversion principle with Depends as dependency injection. I have this:

```Python
# infraestructure/countries.py are my endpoints
@router.get('/')
async def all_countries(all_countries: AllCountries = Depends(AllCountries)) -> List[Country]:
    countries = await all_countries.handle()
    return ResponseModel(countries)
```

```Python
# infraestructure/mongo_country_repository.py 
# is my mongo repository to access to the collection
class MongoCountryRepository(CountryRepository, MongoBaseRepository):
    manager: DatabaseManager
    model: Country
    collection: str

    def __init__(self, manager: DatabaseManager = Depends(get_database)) -> None:
        self.manager = manager
        self.model = Country
        self.collection = 'countries'

    async def get(self, id: str) -> Country:
        return await self.get_record(id)

    async def all(self) -> List[Country]:
        return await self.get_all()

    async def create(self, record: Country) -> Country:
        return await self.insert_record(record)
```

And I have my application service, in another folder (not infrastructure)

```Python
# application/all_countries.py
# is my application service for get the countries
class AllCountries(Service):
    repository: CountryRepository

    def __init__(self, repository: CountryRepository) -> None:
        self.repository = repository

    async def handle(self) -> List[Country]:
        return await self.repository.all()
```

This example not is 100% real, because the logic is very easy, but is for demonstrate my problem.

I would like that the application service know nothing about fastapi, and I'm trying all dependencies manage to in infrastructure.
My question is how can set the repository MongoCountryRepository to AllCountries application service?. CountryRepository is an interface.

**Comment by teamhide** [comment]
Maybe it looks like part of clean architecture. I also had the same concerns before and implement dependency injection library. How about use it? https://github.com/teamhide/pythondi

**Comment by ChilloManiac** [comment]
Would love to be able to accommodate this case in native FastApi

**Comment by tiangolo** [comment, marked as answer]
You probably need a different tool, the dependency injection system in FastAPI is very simple and it's designed to be interested and handled by FastAPI. You probably need something else for your internal dependency system and then combine that with FastAPI.

**Reply by 11kkw** [reply]
Are there any third-party libraries you would recommend for managing complex dependencies? Also, I’m curious if FastAPI has plans to support more advanced dependency injection features in the future.

**Comment by SobikXexe** [comment]
You can just use factory.

```python
def all_countries_factory():
    return AllCountries(repository=MongoCountryRepository())

@router.get('/')
async def all_countries(all_countries: AllCountries = Depends(all_countries_factory)) -> List[Country]:
    countries = await all_countries.handle()
    return ResponseModel(countries)
```
